src/middlewared/middlewared/main.py

```diff
@@ -106,12 +105,27 @@ def register_callback(self, name, method):
         self.__callbacks[name].append(method)
 
     def _send(self, data):
+        serialized = json.dumps(data)
+        asyncio.run_coroutine_threadsafe(self.response.send_str(serialized), loop=self.loop)
+        if sys.getsizeof(serialized) > 1000000:
+            # no reason to store data in the deque that
+            # is larger than 1MB after being serialized.
+            # This gets _really_ painful on systems with
+            # many (100's) of snapshots because running a
+            # simple `zfs.snapshot.query` via midclt from
+            # the cli produces ridiculously large output.
+            # Caching that in the main middleware process
+            # is excessive and only hurts us. Instead we'll
+            # just store a string indicating it was excluded.
+            message = 'RESULT EXCLUDED BECAUSE THE SIZE IS > 1MB'
```

Comment:
Can we store just first megabyte of the message and cut the rest in this case? `message = message[:1000000] + '...'`

Reply:
Great idea, I'll fix.